pymeasure/instruments/common_base.py

```diff
@@ -119,6 +120,49 @@ class CommonBase:
 
     def __init__(self):
         self._special_names = self._setup_special_names()
+        # Add children from ChildDescriptors.
+        for item, value in inspect.getmembers(self.__class__):
+            if isinstance(value, ChildDescriptor):
+                if isinstance(value.id, (list, tuple)):
+                    for id, cls in zip(value.id, value.cls):
+                        self.add_child(cls, id, collection=item, **value.kwargs)
+                else:
+                    self.add_child(value.cls, value.id, collection=item, **value.kwargs)
+
+    @staticmethod
+    def children(id, cls, **kwargs):
+        """Describe children of this parent class.
+
+        The children will be added to the parent instance at instantiation with
+        :meth:`add_child`. The variable name (e.g. :code:`channels`) will be
+        used as the `collection` of the children. You may define the attribute
+        prefix via keyword arguments. Normally, do use :code:`channels` as variable
+        and do leave the prefix at the default `"ch"`.
+
+        .. code::
+
+            channels = CommonBase.children(["A", "B"], ChildClass)
```

Comment:
It works exactly as you propose, only, that `inst.channels` (in my example) will be list (and not a dictionary) and that it will create also the attributes `inst.ch_A`, `inst.ch_B`. The attributes offer the possibility of autocompletion, while the list offers a simple way to loop over all elements (which is, however, not that much more difficult with a dictionary).

`inst.power` will be the PowerChannel instance.

I hope that dual approach (list and attributes) is fine with you.

Reply:
> It works exactly as you propose, only, that `inst.channels` (in my example) will be list (and not a dictionary) and that it will create also the attributes `inst.ch_A`, `inst.ch_B`. The attributes offer the possibility of autocompletion, while the list offers a simple way to loop over all elements (which is, however, not that much more difficult with a dictionary).
> 
> `inst.power` will be the PowerChannel instance.
> 
> I hope that dual approach (list and attributes) is fine with you.

From a first glance this looks good, it keeps much of the previous behaviour, but achieves the desired definition on a class level. And I agree, a *sequence of channels* fits well into a list (so, a *sequence*). I'll give feedback in detail once I've correlated my compare view with where I can actually leave comments :-P

Reply:
You are right, sorry I didn't get it at the beginning. It is a bit unusual for a descriptor implementation (for my taste), but it works.
Couples of final questions:
- Any strong argument against dictionary ?
- For my information, do autocompletion works also with proper descriptors (e.g. control/set/read properties?)

Reply:
> Any strong argument against dictionary ?

A dict would duplicate the information we already have in the names of the attributes (`ch_A`) in the keys of the dict.
The stated reason for why we need the collections at instance level (iirc) was to iterate through them, which a list fulfils. 
If the names are important for access, rather use the attributes (i.e. `instr.ch_A` is easier than `instr.channels['A']`). If you need the names when iterating, use the list elements `[e.id for e in self.channels]`. What do you want to use a dict for?

> For my information, do autocompletion works also with proper descriptors (e.g. control/set/read properties?)

Yes (if you mean the `Instrument.control` factory)

Reply:
> > Any strong argument against dictionary ?
> 
> A dict would duplicate the information we already have in the names of the attributes (`ch_A`) in the keys of the dict. The stated reason for why we need the collections at instance level (iirc) was to iterate through them, which a list fulfils. If the names are important for access, rather use the attributes (i.e. `instr.ch_A` is easier than `instr.channels['A']`). If you need the names when iterating, use the list elements `[e.id for e in self.channels]`. What do you want to use a dict for?
> 
I think this potential duplication costs nothing to us and has the benefit of using the channel designator as index, this will result in something more 'natural' when doing loops.
For example, consider the use case that you want to initialize all the channels and do something special for channel 'B', in this case I would do a loop like this:
```
for designator, value in self.channels.items():
    if designator=='B':
        something
    else
        something 

```
Similarly when channel designator is a number often starting with 1, list will start with index 0.

In summary, you don't have to make any index translation when writing code loops.

If you look from another perspective, python already offer data structures with indexing to handle the case we want to support. Why inventing a new, not pythonic in my opinion, syntax with '_' to implement indexing?

Reply:
Ok. So we exchange the collection lists for collection dictionaries?

Reply:
This is a good example, it can even more easily be done with lists, too:
```python
for channel in self.channels:
    if channel.id == 'B':
        something
    else:
        something
```
I'd argue that it's simpler (no separate `items()` call) and more pythonic (iterating through a sequence of channels as you call it using a sequence datastructure) and more encapsulated/separation of concerns (no potential to desync between that dict key and the corresponding value's `id` attribute, the `id` of a channel should primarily reside with the channel object). 

>  data structures with indexing to handle the case we want to support.

The use case for this is not indexing, though, but iteration! 
Indexed access is handled by `instr.prefix_id` specifically because it's more concise than having to do `instr.channels["id"]`. Also, this offers autocomplete of matching entries in the REPL, which a dictionary does not. 

If these points don't convince you, I don't know what more I can add. :shrug: 
Maybe I am missing something?

> Similarly when channel designator is a number often starting with 1, list will start with index 0.

I'm not sure what you mean? If you channel number starts with 1, your channel's `id` attrs also start with 1? What index do you have to translate?

> this potential duplication costs nothing to us

Duplication always has a cost. 
Zen of Python: *There should be one-- and preferably only one --obvious way to do it.*
Indexed access -> use the instr attributes. Iterating access -> use the list.

> handle the case we want to support.

I think the problem here could be that we are slightly shifting goalposts, here. Initially you requested index access (`inst[1] # same as inst.name1`), and the possibility to iterate through the "children" (e.g. to do analysis). This implementation does that. 
Now, this is also supposed to handle indexed access.

I'd say for the next "big" feature, we should hash out the design in more detail in an issue, before doing the implementation.

Reply:
> This is a good example, it can even more easily be done with lists, too:
> 
> ```python
> for channel in self.channels:
> if channel.id == 'B':
> something
> else:
> something
> ```
> 
> I'd argue that it's simpler (no separate `items()` call) and more pythonic (iterating through a sequence of channels as you call it using a sequence datastructure) and more encapsulated/separation of concerns (no potential to desync between that dict key and the corresponding value's `id` attribute, the `id` of a channel should primarily reside with the channel object).
> 

Objection: my code can be written by anyone who knows python, your code can be written by anyone who knows internal implementation of channels in PyMeasure.

Which method is more natural and obvious for people not familiar with internal implementation?

> > data structures with indexing to handle the case we want to support.
> 
> The use case for this is not indexing, though, but iteration! Indexed access is handled by `instr.prefix_id` specifically because it's more concise than having to do `instr.channels["id"]`. Also, this offers autocomplete of matching entries in the REPL, which a dictionary does not.
> 
Here, I was referring to the general case of channels who are, in my view, iteration of same set of functions with an index to distinguish them.

My point is that you have invented a new method of indexing for the following reasons:
- ch_A is shorter than channels['A'] (obviously why `ch` in one case and `channels` in the other for comparison is interesting)
- REPL does offer autocompletion for dictionary indexes

I am sorry, but these are not valid arguments, in my opinion.

> If these points don't convince you, I don't know what more I can add. 🤷 Maybe I am missing something?
> 

I am sorry for the long discussion, I wouldn't be honest to say they are convincing me, but far enough, we need to break the tie and progress in the implementation.

Just to be 100% sure, do you see any harm in using a dictionary instead of list ?

> > Similarly when channel designator is a number often starting with 1, list will start with index 0.
> 
> I'm not sure what you mean? If you channel number starts with 1, your channel's `id` attrs also start with 1? What index do you have to translate?
> 

If I write a loop similar to the one above and do something special for channel 2, I need to check for index 1 as all indexes are shifted backward by 1. Again confusing in my opinion.

> > this potential duplication costs nothing to us
> 
> Duplication always has a cost. Zen of Python: _There should be one-- and preferably only one --obvious way to do it._ Indexed access -> use the instr attributes. Iterating access -> use the list.
> 

Easy answer here (inline with Zen of Python): use dictionary and you target both.

> > handle the case we want to support.
> 
> I think the problem here could be that we are slightly shifting goalposts, here. Initially you requested index access (`inst[1] # same as inst.name1`), and the possibility to iterate through the "children" (e.g. to do analysis). This implementation does that. Now, this is also supposed to handle indexed access.
> 

Yes, you are right in this. This was part of the initial brainstorming and I think this particular point has been evolved during discussion. I would say this is common during reviews.

> I'd say for the next "big" feature, we should hash out the design in more detail in an issue, before doing the implementation.

I agree, this is a good proposal.

Reply:
> I'd say for the next "big" feature, we should hash out the design in more detail in an issue, before doing the implementation.

I agree. In the beginning, I did not expect it to grow into that much discussion. Just adding the channel and commonbase...

Reply:
#### First about the list vs. dictionary debate:
Dictionaries offer more "natural" access for those who do not know pymeasure's inner working well and those who want to address it via the channel id in a programmatic way (`gettatr` is possible but less known for beginners), but they are no obstacle for looping.
Therefore I'm for a dictionary.

#### Second regarding the additional attribute (`ch_A`):
- I did not know, that REPL supports autocompletion for dictionary keys, thanks for the hint.
- With channel ids, that autocompletion is basically useless, as (at least my REPL) does offer all kinds of different commands like Exceptions, all... as well, such that you do not see the channels
- At least with my (german) keyboard, writing `channels["A"]` is a lot more complicated than `ch_A` due to the square brackets.
- Addressing a channel via attribute and dictionary is almost redundant. For the sake of code homogeneity, I'd forsake that attribute and stick with a dictionary, as some implementors might use `ch_A` and others `channels['A']`

#### Third regarding discussions
- I think it is good, to talk about code and to try the implementation
- On the other hand, finding that particular fundamental discussion in the code review, is a hassle

What do you think about having a discussion and a PR (or branch) as a pair (for larger changes)?
That way you can talk about the code, but also have a more ordered place to talk about implementation philosophy. In the discussion you can have one comment for every topic (e.g. one for the Creator, one for the list vs dict implementation etc.) and then you reply to the specific topic. An issue makes it more difficult to separate those different topics.
The branch/PR offers code to talk about.

Finally in the PR itself you talk about concrete coding/wording issues.

Reply:
> internal implementation of channels in PyMeasure.

`Channel.id` is a public attribute, created by default, though? What do you mean with "internal"?

> REPL does offer autocompletion for dictionary indexes

Using plain python (3.9.9) dict members (so, class attributes) get autocompletion, dictionary entries do not:
```
>>> dummy = {'a': 5, 'b': 10}
>>> dummy.<tab>
dummy.clear(       dummy.fromkeys(    dummy.items(       dummy.pop(         dummy.setdefault(  dummy.values(      
dummy.copy(        dummy.get(         dummy.keys(        dummy.popitem(     dummy.update(      
>>> dummy[      # <tab> only generates a tab character
```
So, I'm assuming you mean "IPython" when you says REPL?

> If I write a loop similar to the one above and do something special for channel 2, I need to check for index 1 as all indexes are shifted backward by 1. Again confusing in my opinion.

No, because channel 2 will have `.id==2`, as I understand it?

> Just to be 100% sure, do you see any harm in using a dictionary instead of list ?

Yes, otherwise I would not argue against it vehemently. :-D

> I am sorry for the long discussion, I wouldn't be honest to say they are convincing me, but far enough, we need to break the tie and progress in the implementation.

Yes, I agree, we are at a stalemate.

At this point, I'm prepared to call it quits and give up. Please resolve this part of the feature as you both consider best. You know my concerns and I trust you to keep them in mind as well as possible.

Reply:
Let's continue the discussion on a better implementation flow in the main thread, so it doesn't get lost here.

Reply:
At first I did not see a great difference between list and dict, but from this

> I'd argue that it's simpler (no separate items() call) and more pythonic (iterating through a sequence of channels as you call it using a sequence datastructure) and more encapsulated/separation of concerns (no potential to desync between that dict key and the corresponding value's id attribute, the id of a channel should primarily reside with the channel object).

I extracted a very important argument: the channel's `id` might get different from the dictionary key (either because the channel id changed or because the key changed. Right now, I use the id to remove a dictionary entry.

Do we expect people to change either id or key? I see the case, that someone might change the id, because he wants to use the channel for something else.

What do you think about making the channel `id` a protected attribute (by using a property)? That way we exclude one desync possibility.

Reply:
Do you then also want to protect the dictionary keys against change?

BTW, just realised, many instruments already have an ` id` property, which is unconnected to a Channel's `id` (attribute/property) -- should we choose a different name for the latter to disambiguate?

Reply:
>  Do you then also want to protect the dictionary keys against change?

I don't see a straight forward possibility. I, personally, just want to enable channels and try to make some implementation with which everyone can live. I see the advantages of the list (and some of the dictionary), but can live with either one. It is frustrating, that the whole work hangs due to one point of disagreement.

>  BTW, just realised, many instruments already have an id property, which is unconnected to a Channel's id (attribute/property) -- should we choose a different name for the latter to disambiguate?

If we called it name, we could add it directly to the CommonBase. In the beginning you did not want to use that name, as it is already used in the instrument, but maybe that's not that bad.

Reply:
> If we called it name, we could add it directly to the CommonBase. In the beginning you did not want to use that name, as it is already used in the instrument, but maybe it's good, as it defines the channels and the instrument.

Yeah, the namespace is getting crowded :D Yeah, `name` still collides (in the signature). `id` is still best, as it avoids that signature ambiguity, I just wanted to throw this out there in case one of you had a good idea! Let's leave `id`.

Reply:
> > Do you then also want to protect the dictionary keys against change?
> 
> I don't see a straight forward possibility. 

I agree. Maybe the question was in a socratic sense, to demonstrate how some choices create "friction" later on.

> I, personally, just want to enable channels and try to make some implementation with which everyone can live. I see the advantages of the list (and some of the dictionary), but can live with either one. It is frustrating, that the whole work hangs due to one point of disagreement.

Yes, this whole thing went down suboptimally. This is probably one challenge in an asynchronous, fully distributed and remote operation, otherwise we could have just sat together in a meeting room for an afternoon, put our heads down and hashed out the design.

> At this point, I'm exhausted and prepared to call it quits and give up. Please resolve this part of the feature as you both consider best. You know my general/other concerns regarding design and I trust you to keep them in mind as much as possible.

Maybe this was not clearly enough formulated by me, from my point of view I backed down and let you two do this your way. It doesn't make sense to hold this up further, I have made my case and could not convince you. No hard feelings :peace_symbol:

Reply:
I am sorry I lost the thread, why would we want to protect dictionary keys against change ?

Reply:
> I am sorry I lost the thread, why would we want to protect dictionary keys against change ?

I tried to mitigate one difficulty with a dictionary, that it's key might get out of sync with the ID, in order to find a solution fit for everyone.

> Maybe this was not clearly enough formulated by me, from my point of view I backed down and let you two do this your way. It doesn't make sense to hold this up further, I have made my case and could not convince you. No hard feelings ☮️ 

I understood you, but I did not want to continue against your strong points as if you had never raised them. 

As you @bilderbuchi backed down, I do not see any sensitive alternative to dict or list, and @msmttchr approves dictionaries, I resolve this conversation.